index.coffee

```diff
@@ -178,10 +187,34 @@ class ResumableRequest extends stream.Readable
 			@error = new Error('Received more bytes than expected!')
 		else if @retries >= @maxRetries
 			@error = new Error('Maximum retries exceeded')
+		else if @bytesTotal? and @bytesRead is @bytesTotal
+			@_end()
+			return
 		else
 			@retries += 1
 			@emit('retry', @progress())
 			@_request()
 			return
 
-		@abort()
+		@_abort(@error)
+
+	abort: ->
+		@request?.abort()
+
+	_abort: (err) ->
+		return if @ended
+		@ended = true
+
+		@response?.resume()
+		@response?.destroy()
+
+		if err?
+			@emit('error', err)
+		@emit('aborted')
+		@emit('end')
+
+	_end: ->
+		return if @ended
+		@ended = true
+		@response.destroy()
+		@emit('end')
```

Comment:
There shouldn't be any need to manually emulate an "end" event – the `_read()` method (or any other method that would cause ending of the stream, like i.e. `_retry()`) should `push(null)` once the source has ended, which will automatically end the stream. Through that the stream will also properly error in write-after-end situations.